### ISSUE TITLE
Linux Client: Multiple bugs in idle detection

### ISSUE THREAD

**Issue by RichardHaselgrove** [BOINC/boinc#3715]
**Describe the bug**
Not my personal observation, but drawing attention to a detailed error report and workround in
https://boinc.berkeley.edu/forum_thread.php?id=13673
Multiple users reporting.

**Steps To Reproduce**
1. Problem reported in v7.16.6 release
2. Problem resolved by downgrading to v7.16.3

**Expected behavior**
Return to previous behavior - suspend when user is active, resume computation when user is inactive.

**System Information**
 - OS: Ubuntu/Kubuntu 20.04 (change from behavior in 19.10)
 - BOINC Version: 7.16.6 (problem exists), 7.16.3 (problem resolved)

**Additional context**
We should have a change list somewhere so that we can identify the exact changes between v7.16.3 and v7.16.6, where this might have been introduced.

**Comment by Germano0** [comment]
This is no longer a problem on Fedora / CentOS / RHEL due my patch
https://src.fedoraproject.org/rpms/boinc-client/blob/master/f/disable_idle_time_detection.patch

**Comment by RichardHaselgrove** [comment]
Unfortunately, simply smothering the bug kills the intended behaviour (noise reduction, freeing resources for paying work) too.

**Comment by mpsdskd** [comment]
Then I could also just disable idle detection in the GUI and have the same effect?

> This is no longer a problem on Fedora / CentOS / RHEL due my patch
> https://src.fedoraproject.org/rpms/boinc-client/blob/master/f/disable_idle_time_detection.patch
> Note: I have tested (again) BOINC idle time detection before deciding to remove all the code

**Comment by RichardHaselgrove** [comment]
Better that we find what stopped it working and restore the intended behaviour.

**Comment by Germano0** [comment]
Actually I was remembering it wrong
https://bugzilla.redhat.com/show_bug.cgi?id=1822723#c2

**Comment by mpsdskd** [comment]
In the boinc forum thread, one user also references the following _folding@home_ forum threads:

https://foldingforum.org/viewtopic.php?f=89&t=33351
https://foldingforum.org/viewtopic.php?f=89&t=33898

In the first one there is a workaround script, which controls F@H using xprintidle - might also work for boinc, but is also at best a temporary workaround.

**Comment by Agbar** [comment]
Appears to be fixed in 8a1099f816652cffdd24d8d0c91b36e6ae7ef03. It is available since [client_release/7.16/7.16.8](https://github.com/BOINC/boinc/releases/tag/client_release/7.16/7.16.8) but not officially released.

**Comment by AenBleidd** [comment]
@Agbar, did you test it with local build?
@RichardHaselgrove, are you able to test this? Linux binaries could be retrieved from BinTray

**Comment by Agbar** [comment]
@AenBleidd, yes, I tested it with a local build.
Eyeballing the code on question reveals that before the fix it always returned 0 [seconds of inactivity]. It was taking min (0, user_idle_time) which always evaluates to 0. And 0 means you are moving the mouse right now.
Fixed code calculates min(user_idle_time, infinity), I believe it is correct.

**Comment by RichardHaselgrove** [comment]
@AenBleidd - posting from my Linux machine with v7.16.11, but sadly I can't help - I've never suffered this problem (older version of Linux), and as I said in opening the issue, I'm merely the messenger for other users.

@Agbar's analysis of the code sounds like a horribly plausible bug.

**Comment by mauromol** [comment]
Same issue here with 7.16.6 under Kubuntu 20.04, worked fine with the older version (I can't remember what it was) under Linux Mint 18.3 (based on Ubuntu 16.04).
It's not clear to me whether an updated package can be installed in some way...

Not good to see that on the official BOINC website the recommended version is 7.4.22, with 7.16.6 being "experimental", though Ubuntu LTS packages supply the experimental version...

**Comment by pilot51** [comment]
Same here on Linux Mint 20 with 7.16.6. I first noticed it a few months ago and just dealt with it by switching the option in preferences if I happened to remember. Just now I noticed all my tasks expired because I forgot for a long time, so I searched for others with the issue or a solution and found this.

Since it sounds like the fix is releasing soon, I'll wait for it since I don't think I want to uninstall to force 7.16.3 with potential side effects.

**Comment by RichardHaselgrove** [comment]
OK, conducted a small test. This machine shows up on projects as
`Linux Mint 19.1 Tessa [4.15.0-115-generic|libc 2.27 (Ubuntu GLIBC 2.27-3ubuntu1.2)]`
BOINC is v7.16.11 from @LocutusOfBorg's PPA

I normally set machine preferences to run always, not to suspend when in use. Reversed both of those - and it carried on running as if nothing had changed.

Enabled 'idle_detection_debug' logging, and got
```
Sun 11 Oct 2020 15:43:28 BST |  | [idle_detection] Error (2) opening /tmp/.X11-unix/.
Sun 11 Oct 2020 15:43:28 BST |  | [idle_detection] Could not connect to any DISPLAYs. XSS idle determination impossible.
```
repeated once per second.

This machine is currently running with an elderly PS/2 keyboard and mouse (I ran out of working USB peripherals). I'll repeat later with the machine connected to my USB KVM.

**Comment by RichardHaselgrove** [comment]
There is a 'socket file' called X0 in /tmp/.X11-unix/

**Comment by RichardHaselgrove** [comment]
USB peripherals make no difference. Same versions of BOINC and Linux, same X0 socket, same result, same error message. Both machines have full HD graphical desktop, driven from NVidia GTX 1660 GPUs.

**Comment by comperem** [comment]
I've got this issue as well. Boinc Manager 7.16.6 (x64) fails to restart worker processes when machine is idle for more than 3 minutes (or many hours). Issue demonstrated on Ubuntu 20.04.1.

**Comment by siriak** [comment]
Any ideas on when a fixed version (7.16.8?) will be available through apt?

**Comment by AenBleidd** [comment]
@siriak, for Windows the latest released version is 7.16.11 that can be installed via this PPA:
https://launchpad.net/~costamagnagianfranco/+archive/ubuntu/boinc

**Comment by AenBleidd** [comment]
Should be fixed via #3512

**Comment by mauromol** [comment]
> @siriak, for Windows the latest released version is 7.16.11 that can be installed for linux via this PPA:
> https://launchpad.net/~costamagnagianfranco/+archive/ubuntu/boinc

I guess you meant Ubuntu here?

Meanwhile I opened https://bugs.launchpad.net/ubuntu/+source/boinc/+bug/1900803
I think Ubuntu should better port the fix to their official repos as well...

**Comment by RichardHaselgrove** [comment]
@AenBleidd: No. #3512 fixes #3509, which is a different problem. The sense is opposite: #3509 is about BOINC not re-starting: this issue is about BOINC not stopping. This one is not fixed, even in v7.16.11. Reopening.

**Comment by AenBleidd** [comment]
@RichardHaselgrove, yes, you're right. Thanks for this

**Comment by mauromol** [comment]
By the way, I tried to install BOINC client and Manager 7.16.15 from https://bugs.launchpad.net/ubuntu/+source/boinc/+bug/1900803 in my Kubuntu 20.04 to check whether this bug was fixed, but I get error "gui_rpc_auth.cfg exists but can't be read. Check file permissions" error on manager start and I don't see my project settings (BOINC manager is completely empty). Previously I had installed boinc-client and boinc-manager from the official Ubuntu repos. Permissions are:

- in /var/lib/boinc-client:
  lrwxrwxrwx  1 root  root      34 ott 21 09:28 gui_rpc_auth.cfg -> /etc/boinc-client/gui_rpc_auth.cfg
- in /etc/boinc-client:
  -rw-r-----   1 root  boinc     1 feb 23  2016 gui_rpc_auth.cfg

I tried to chmod g+w on /etc/boinc-client/gui_rpc_auth.cfg but the error persists. Any suggestion?

**Comment by AenBleidd** [comment]
@mauromol, please try this: https://boinc.berkeley.edu/dev/forum_thread.php?id=14013#101083

**Comment by RichardHaselgrove** [comment]
And also follow https://boinc.berkeley.edu/dev/forum_thread.php?id=14013&postid=101191#101191

**Comment by mauromol** [comment]
Thanks @AenBleidd and @RichardHaselgrove , I fixed my problem and now I can run the new BOINC version.
But then... do I am right to say that the version 7.16.15 of the BOINC client from https://launchpad.net/~costamagnagianfranco/+archive/ubuntu/boinc is now just simply ignoring the setting "suspend when the computer is in use"? I see my computations now running even when I'm actively using my PC, even if they should be suspended until 1 minute idle...

**Comment by RichardHaselgrove** [comment]
That's why I opened (and then reopened) this ticket.

What output do you get from idle_detection_debug (Event Log, from Ctrl+Shift+F)?

**Comment by mauromol** [comment]
@RichardHaselgrove 
I'm a bit lost, this bug was about BOINC being always in suspended state ("computer is in use"), if now it has become "computer is never in use" (hence BOINC is never suspended) the issue title should probably be changed...

Anyway, my event log is full of these:

    mer 21 ott 2020, 19:37:33 |  | [idle_detection] Error (2) opening /tmp/.X11-unix/.
    mer 21 ott 2020, 19:37:33 |  | [idle_detection] Could not connect to any DISPLAYs. XSS idle determination impossible.

**Comment by RichardHaselgrove** [comment]
Yup, that's what I get too. At least it's consistent.

**Comment by RichardHaselgrove** [comment]
Took @mauromol's advice, and made a collective title for the whole range of problems.

Working with 

> Sun 11 Oct 2020 15:43:28 BST |  | [idle_detection] Error (2) opening /tmp/.X11-unix/.

"Error 2" is "#define ENOENT 2 /* No such file or directory */" - but the file is there. I can see it, even if BOINC can't.

I made a sub-folder in the boinc-client tree, made an empty file with the same name, compiled a client with the changed folder location string in https://github.com/BOINC/boinc/blob/master/client/hostinfo_unix.cpp#L1819, and idle detection now works.

tl;dr BOINC can't read special system files.

**Comment by comperem** [comment]
ok, so what's the current status. should I uninstall and reinstall Boinc? I'm not interested in compiling from source or detailed troubleshooting other than uninstall/install.

My Boinc will not turn back on when "Suspend when computer is in use" is enabled.

Is this fixed?

**Comment by RichardHaselgrove** [comment]
@comperem: Looking through this tread and the related code, it looks as if **that particular bug** (won't turn back on) was:

created on 30 October 2019 (6d1d44b5740255a3286da017b1f2d3b0093a350f)
fixed on 22 March 2020 (8a1099f816652cffdd24d8d0c91b36e6ae7ef031)

The bug was transferred to the release branch for the first time at release tag 7.16.5, and corrected in the release branch at tag 7.16.8.

So, again for **that particular bug**, avoid minor versions .5, .6, and .7 - re-installing with any version **apart from those** should improve things.

**Comment by comperem** [comment]
Thank you Richard. I cleaned up my Ubuntu 20.04.1 ppa list and uninstalled boinc-client and boinc-manager but was still getting only the 7.16.6 version that still has the problematic issue.

That's because the [Ubuntu repositories](https://launchpad.net/ubuntu/+source/boinc) still have the 7.16.6 version.

But after adding the Boinc development release (which seems to still be stable and good), this caused the next `apt-get update ; apt-get upgrade` to install the Boinc 7.16.15.

I think it's likely to be fixed now. I'll be watching. Thanks for the clarity and discussion on this fix.

**Comment by comperem** [comment]
there were some other issues causing problems besides the PPA.

After getting the 7.16.15 version installed it didn't start up correctly. [this was the issue and how it was resolved](https://boinc.berkeley.edu/dev/forum_thread.php?id=14023&postid=101360#101360), at least in Ubuntu 20.04.1 that was upgraded from Ubuntu 18.04.5.

**Comment by pilot51** [comment]
I used [ppa:costamagnagianfranco/boinc](https://launchpad.net/~costamagnagianfranco/+archive/ubuntu/boinc) to update from 7.16.6 to 7.16.15 and the behavior has flipped. Instead of BOINC thinking the computer is always in use, it now thinks the computer is never in use and continues running even though I have "Suspend when computer is in use" enabled and "Run based on preferences" selected.

**Comment by RichardHaselgrove** [comment]
If you turn on <idle_detection_debug> event logging, does it report
```
Sun 11 Oct 2020 15:43:28 BST |  | [idle_detection] Error (2) opening /tmp/.X11-unix/.
Sun 11 Oct 2020 15:43:28 BST |  | [idle_detection] Could not connect to any DISPLAYs. XSS idle determination impossible.
```
as reported (twice) above? If so, we know what the problem is (BOINC can't read /tmp/.X11-unix/) - but I can read mine. We're waiting for a Linux guru to come forward and point us to a solution.

**Comment by pilot51** [comment]
Yes, same thing once or twice every second.

When I ran `sudo nemo /tmp` to try changing the folder access permission to see if that helps, BOINC suddenly detected the computer in use. I did not actually change the permission, just opened Nemo as superuser.

**Comment by RichardHaselgrove** [comment]
That sounds like a clue... ;-)

**Comment by RichardHaselgrove** [comment]
Hmmm. Didn't work for me - I'm running an old version of Mint, so I had to try with caja. I also have some big tasks running, which don't ckeckpoint very often. But the error message repeats even with 'Activity run always', so I kept that visible - didn't stop when I opened caja (sudo or not, /tmp or root, gui or terminal).

As to the cause - the failure is within code introduced by 69abe830e6069ec2139783392e8d3a73fb7e6baf (30 Dec 2015). The current version still contains the comment
```
    // ... If we can't open this
    // directory, or the display_values vector is otherwise empty, then a
    // static list of guesses for open display servers is utilized instead
    // (DISPLAY values ":{0..6}") that will attempt connections to the first
    // seven open Xservers.
```
Although the comment is still there, the code isn't. It was removed by 7ff020c48781099c02c71020e2518a0bd13c65c3 (23 Apr 2020), part of #3549 and referring to issue #2256. Did idle detection ever work, say on builds before #3549?

@BryanQuigley - are you still reading here, and do you have any ideas?

**Comment by BryanQuigley** [comment]
Note the costamagnagianfranco ppa and Debian builds are  7.16.15+dfsg.is.7.16.11 so really 7.16.11 (which I also missed).

Quick guess (not based on any version checking):

sudo systemctl edit boinc-client.service
[Service]
PrivateTmp=false

Then reload and restart the service.  If that is it, should be an easy item to get upstream.

**Comment by pilot51** [comment]
Long story short, BOINC started running again after exactly 5 minutes. I changed the 'in use' timeout to 0.2 minutes to speed up testing and found that it suspends when I run any sudo command (such as `sudo nano` or `sudo -i`) and enter the first character in the password; actually logging in is not required. I tried logging into Synaptic and that didn't trigger detection. I tried the same trick with `su` and that didn't work. The errors continued being logged throughout all these tests.

**Comment by BryanQuigley** [comment]
@pilot51 quick take - the command line could cause tty idle detection to be triggered

**Comment by pilot51** [comment]
`PrivateTmp=false` worked. Thanks!
I noticed that it mysteriously detects activity about 15-30 seconds after I stop touching the keyboard or mouse. That's a separate issue, perhaps caused by another application, and doesn't appear to be a problem with a more reasonable timeout.

**Comment by RichardHaselgrove** [comment]
Added `PrivateTmp=false` per instructions. Didn't want to keep interrupting my long tasks, but the error message changed to
```
Sun 15 Nov 2020 18:53:37 GMT |  | [idle_detection] XSS idle detection succeeded on display ':0'.
Sun 15 Nov 2020 18:53:37 GMT |  | [idle_detection] display idle time: 0 sec
```
I think we'll take that as a success. @smoe, @LocutusOfBorg - could we prevail on you for one more test build, please?

**Comment by LocutusOfBorg** [comment]
ack, uploaded will go in some minutes.

**Comment by RichardHaselgrove** [comment]
Hmm. You win some, you lose some. I've got

`PrivateTmp=false`

and
```
Mon 16 Nov 2020 18:09:53 GMT |  | [idle_detection] XSS idle detection succeeded on display ':0'.
Mon 16 Nov 2020 18:09:53 GMT |  | [idle_detection] display idle time: 15 sec
```
but boinc is still running. I'll think about that, and check settings.

Ah - when I set both Activity menu items to 'according to preferences', the GPU tasks pause, but the CPU tasks don't - we're half way there!

**Comment by RichardHaselgrove** [comment]
Sorry, missed a checkbox. They both work. Success!

**Comment by smoe** [comment]
Thank you for your efforts!

**Comment by mauromol** [comment]
So, what's the final state of this? I mean:

1. with which minimal version the following configuration change is needed to completely fix the problem?

```
sudo systemctl edit boinc-client.service
[Service]
PrivateTmp=false
```
2. which versions are affected by this problem for sure, even if you change the above configuration? In particular, is the 7.16.6 version packaged for Ubuntu 20.04 still broken even if you apply this change?
3. does a newer version shipping this configuration change by default already exist?

Thanks in advance!

**Comment by RichardHaselgrove** [comment]
This isn't a BOINC version issue - it concerns the independent downstream packaging performed by each independent Linux distro management team. You'll have to ask them when their next release cycle is due - that's not under BOINC's control.

**Comment by mauromol** [comment]
Hmmm... so wasn't there any BOINC bug at all? Honestly I didn't understand this from the above discussion...

**Comment by LocutusOfBorg** [comment]
@mauromol please open a bug against ubuntu boinc package and link it here, so I'll SRU the new version there.

**Comment by LocutusOfBorg** [comment]
https://bugs.launchpad.net/ubuntu/+source/boinc/+filebug

**Comment by mauromol** [comment]
> @mauromol please open a bug against ubuntu boinc package and link it here, so I'll SRU the new version there.

@LocutusOfBorg  I already opened https://bugs.launchpad.net/ubuntu/+source/boinc/+bug/1900803

**Comment by RichardHaselgrove** [comment]
I think we've now successfully resolved the two major issues under discussion:

Problem 1: BOINC won't start (permanently 'in use') - bug in versions 7.16.5/6/7. Avoid those versions. Fixed in #3512
Problem 2: BOINC won't stop (doesn't react to mouse/keyboard use) - BOINC can't read the X-server names. Needs a Linux guru. Fixed in #4105 

We may have some follow-up work to do in #4125, but I think we can finally close this one.

**Comment by original-birdman** [comment]
> This isn't a BOINC version issue - it concerns the independent downstream packaging performed by each independent Linux distro management team. You'll have to ask them when their next release cycle is due - that's not under BOINC's control.

Whereas that may be true, I build my own BOINC from the BOINC sources (historically - to ensure I have the same release version across multiple different distributions - one of which is Entware on an Asus router).

So knowing which BOINC release these fixes are in on got *is* of use to some of us.

**Comment by RichardHaselgrove** [comment]
@original-birdman: Fair comment. It is in the [job description](https://github.com/BOINC/boinc-policy/blob/master/Development_Documents/Client_Release_Process.md) for the Release Manager to provide (or delegate the provision of) a change list for each released version. Just at the moment, we are in a difficult position, because there have been multiple attempts to release a complete working version of BOINC v7.16 across all supported platforms. I hope we can reach that stage soon.
